### PR TITLE
SILOptimizer: use BasicBlockSet instead of SmallPtrSet in various transformations.

### PR DESCRIPTION
**Pull request by eeckstein**
BasicBlockSet is more efficient than a SmallPtrSet<SILBasicBlock *>. This change reduces the number of memory allocations and therefore compile time.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci smoke test macOS

**Comment by eeckstein** [comment]
@swift-ci test

File: include/swift/SIL/SILBitfield.h [hunk 1]
```diff
@@ -161,7 +164,45 @@ class BasicBlockSet {
...
+/// An implementation of `llvm::SetVector<SILBasicBlock *,
+///                                       SmallVector<SILBasicBlock *, N>,
+///                                       BasicBlockSet>`.
+///
+/// Unfortunately it's not possible to use `llvm::SetVector` directly because
+/// the BasicBlockSet constructor needs a `SILFunction` argument.
+///
+/// Note: This class does not provide a `remove` method intentinally, because
+/// it would have a O(n) complexity.
+template <unsigned N> class BasicBlockSetVector {
```
Comment (gottesmm) [inline include/swift/SIL/SILBitfield.h:179]: Can you change the name of this to reflect that one can only add and can never remove? The name suggests it is a general utility.

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 2]
```diff
@@ -416,6 +416,17 @@ void JointPostDominanceSetComputer::findJointPostDominatingSet(
...
 
+  /// A set that guards our worklist. Any block before it is added to worklist
+  /// should be checked against visitedBlocks.
+  SILFunction *function = dominatingBlock->getParent();
+  BasicBlockSet visitedBlocks(function);
+
+  /// The set of blocks where we begin our walk.
+  BasicBlockSet initialBlocks(function);
+
+  /// The set of blocks where we begin our walk.
```
Comment (gottesmm) [inline lib/SIL/Utils/BasicBlockUtils.cpp:427]: I think I typed this. Can you fix the comment?

Comment (gottesmm) [reply lib/SIL/Utils/BasicBlockUtils.cpp:427]: Also, why is isLeakingBlock a BasicBlockFlag? I would assume that this would be a BasicBlockSet? Can you elaborate on when I should use one or the other?

File: lib/SILOptimizer/Transforms/DeadCodeElimination.cpp [hunk 3]
```diff
@@ -804,8 +770,42 @@ void DCE::markControllingTerminatorsLive(SILBasicBlock *Block) {
...
+    if (!PDT->getRootNode())
+      return;
+
+    DCE dce(F, PDT);
+    if (dce.run()) {
+      using InvalidationKind = SILAnalysis::InvalidationKind;
+      unsigned Inv = InvalidationKind::Instructions;
+      if (dce.mustInvalidateCalls())
+        Inv |= (unsigned)InvalidationKind::Calls;
+      if (dce.mustInavlidateBranches()) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/DeadCodeElimination.cpp:798]: Typo.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp [hunk 4]
```diff
@@ -356,12 +360,10 @@ class ArrayPropertiesAnalysis {
...
   }
 
-  SmallPtrSetImpl<SILBasicBlock *> &getLoopExitingBlocks() {
+  SmallVectorImpl<SILBasicBlock *> &getLoopExitingBlocks() {
     if (!CachedExitingBlocks.empty())
       return CachedExitingBlocks;
-    SmallVector<SILBasicBlock *, 16> ExitingBlocks;
-    Loop->getExitingBlocks(ExitingBlocks);
-    CachedExitingBlocks.insert(ExitingBlocks.begin(), ExitingBlocks.end());
+    Loop->getExitingBlocks(CachedExitingBlocks);
```
Comment (gottesmm) [inline lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp:366]: Question. Do you know why we used a SmallPtrSet here? Also, I imagine that this can be returned as an ArrayRef.

File: lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp [hunk 5]
```diff
@@ -79,20 +80,20 @@ class UnreachableUserCodeReportingState {
...
+  llvm::SetVector<SILBasicBlock*> PossiblyUnreachableBlocks;
 
   /// The set of blocks in which we reported unreachable code errors.
   /// These are used to ensure that we don't issue duplicate reports.
   ///
   /// Note, this set is different from the PossiblyUnreachableBlocks as these
   /// are the blocks that do contain user code and they might not be immediate
   /// successors of a folded branch.
-  llvm::SmallPtrSet<const SILBasicBlock*, 2> BlocksWithErrors;
+  llvm::SmallPtrSet<SILBasicBlock*, 2> BlocksWithErrors;
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp:91]: Seems like maybe you forgot to finish fixing this part? Shouldn't this be a BasicBlockSet?

File: lib/SILOptimizer/Mandatory/OSLogOptimization.cpp [hunk 6]
```diff
@@ -1402,7 +1403,7 @@ static SILInstruction *beginOfInterpolation(ApplyInst *oslogInit) {
   // formatting and privacy options are literals, all candidate instructions
   // must be in the same basic block. But, this code doesn't rely on that
   // assumption.
-  SmallPtrSet<SILBasicBlock *, 4> candidateBBs;
+  BasicBlockSetVector<4> candidateBBs(oslogInit->getFunction());
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/OSLogOptimization.cpp:1406]: Why is this specific one a BasicBlockSetVector instead of just a BasicBlockSet?

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 7]
```diff
@@ -465,25 +477,28 @@ void JointPostDominanceSetComputer::findJointPostDominatingSet(
...
+      if (visitedBlocks.insert(predBlock))
         worklist.push_back(predBlock);
     }
   }
 
   // After our worklist has emptied, any blocks left in
   // blocksThatLeakIfNeverVisited are "leaking blocks".
-  for (auto *leakingBlock : blocksThatLeakIfNeverVisited)
-    foundJointPostDomSetCompletionBlocks(leakingBlock);
+  for (auto *leakingBlock : blocksThatLeakIfNeverVisited) {
```
Comment (gottesmm) [inline lib/SIL/Utils/BasicBlockUtils.cpp:498]: I don't think this is sufficient. How do you handle if we have the same successor block that must be visited for two different blocks. Then we would call foundJointPostDomSetCompletion with the same block twice which breaks an API guarantee.

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 4 above)
Comment (gottesmm) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp:366]: I haven't thought about it, I am just double checking, I don't think our loop can have multiple of the same blocks passed here, right? Maybe we need an assert to that effect? Your thoughts?

File: lib/SILOptimizer/Transforms/AllocBoxToStack.cpp [hunk 8]
```diff
@@ -142,8 +143,9 @@ static bool addLastRelease(SILValue V, SILBasicBlock *BB,
 // loop.
 static bool getFinalReleases(SILValue Box,
                              SmallVectorImpl<SILInstruction *> &Releases) {
-  SmallPtrSet<SILBasicBlock *, 16> LiveIn;
-  SmallPtrSet<SILBasicBlock *, 16> UseBlocks;
+  SILFunction *function = Box->getFunction();
+  BasicBlockSetVector<16> LiveIn(function);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:147]: Why did you make this semantic change? I guess we were iterating over a SmallPtrSet? So you switched to a BasicBlockSetVector?

**Review by gottesmm — COMMENTED** [review]
Some questions/feedback. Once this compiles, I'll do another look though.

File: include/swift/SIL/SILBitfield.h (hunk 1 above)
Comment (eeckstein) [reply include/swift/SIL/SILBitfield.h:179]: Hm, I like this name. It describes best what this class is.
I even think that there shouldn't be a `remove` method in llvm:SetVector. It is O(n), but people use it as if it was O(0).

File: lib/SIL/Utils/BasicBlockUtils.cpp (hunk 2 above)
Comment (eeckstein) [reply lib/SIL/Utils/BasicBlockUtils.cpp:427]: It's just about naming. BasicBlockFlag is the same as BasicBlockSet, just with a different API. For something like `isLeakingBlock` it's natural to call it a "flag". For something like `visitedBlocks`, it's natural to call it a "set".

Comment (eeckstein) [reply lib/SIL/Utils/BasicBlockUtils.cpp:427]: > Can you fix the comment?

sure

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 4 above)
Comment (eeckstein) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp:366]: >  Do you know why we used a SmallPtrSet here?

I think this was just a bug.

>  I imagine that this can be returned as an ArrayRef.

true. I'll change it.

> I don't think our loop can have multiple of the same blocks passed here, right?

yes

> Maybe we need an assert to that effect? Your thoughts?

I think an assert is not needed (it would be not trivial).

File: lib/SILOptimizer/Mandatory/OSLogOptimization.cpp (hunk 6 above)
Comment (eeckstein) [reply lib/SILOptimizer/Mandatory/OSLogOptimization.cpp:1406]: Because `size()` and `begin()` is used. But I rewrite the code a bit so that BasicBlockSet can be used.

File: lib/SILOptimizer/Transforms/AllocBoxToStack.cpp (hunk 8 above)
Comment (eeckstein) [reply lib/SILOptimizer/Transforms/AllocBoxToStack.cpp:147]: yes, exactly

File: lib/SIL/Utils/BasicBlockUtils.cpp (hunk 7 above)
Comment (eeckstein) [reply lib/SIL/Utils/BasicBlockUtils.cpp:498]: you are right, good catch!
I'll fix it.

File: lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp (hunk 5 above)
Comment (eeckstein) [reply lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp]: Indeed. I changed it

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci smoke test macOS

File: lib/SILOptimizer/Mandatory/OSLogOptimization.cpp (hunk 6 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/OSLogOptimization.cpp]: Its ok. Can you at least add a comment to that effect with a fixme.

File: lib/SIL/Utils/BasicBlockUtils.cpp (hunk 7 above)
Comment (gottesmm) [reply lib/SIL/Utils/BasicBlockUtils.cpp:494]: Can you fix the comment here? I think the comment needs to be updated given functional changes.

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 9]
```diff
@@ -416,6 +416,17 @@ void JointPostDominanceSetComputer::findJointPostDominatingSet(
...
+  /// A set that guards our worklist. Any block before it is added to worklist
+  /// should be checked against visitedBlocks.
+  SILFunction *function = dominatingBlock->getParent();
+  BasicBlockSet visitedBlocks(function);
+
+  /// The set of blocks where we begin our walk.
+  BasicBlockSet initialBlocks(function);
+
+  /// True for blocks which are in blocksThatLeakIfNeverVisited.
+  BasicBlockFlag isLeakingBlock(function);
```
Comment (gottesmm) [inline lib/SIL/Utils/BasicBlockUtils.cpp:428]: Now that we are no longer removing elements from blocksThatLeakIfNeverVisited, why not change that to be a BasicBlockSetVector? Then we can eliminate this set and can just use blocksThatLeakIfNeverVisited, right?

File: lib/SILOptimizer/Transforms/DeadCodeElimination.cpp [hunk 10]
```diff
@@ -450,7 +415,8 @@ void DCE::propagateLiveness(SILInstruction *I) {
 SILBasicBlock *DCE::nearestUsefulPostDominator(SILBasicBlock *Block) {
   // Find the nearest post-dominator that has useful instructions.
   auto *PostDomNode = PDT->getNode(Block)->getIDom();
-  while (PostDomNode && !LiveBlocks.count(PostDomNode->getBlock()))
+  while (PostDomNode && (!PostDomNode->getBlock() ||
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/DeadCodeElimination.cpp:418]: Why was this changed? I am not saying it is incorrect. It just needs a comment at least explaining what the change is.

File: lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp [hunk 11]
```diff
@@ -392,24 +401,24 @@ class YieldOnceCheck : public SILFunctionTransform {
 
       // Find all transitive predecessors of 'yieldPred' that were visited
       // during the analysis
-      SmallPtrSet<SILBasicBlock *, 8> predecessorsOfYieldPred;
+      BasicBlockSet predecessorsOfYieldPred(&fun);
       for (auto *predBB :
            llvm::breadth_first<llvm::Inverse<SILBasicBlock *>>(yieldPred)) {
-        if (visitedBBs.count(predBB)) {
+        if (visitedBBs[predBB].isVisited()) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:407]: One thing that I am noticing in this change is that a bunch of const code is being replaced with non-const code like this. My worry is that we are relying on BasicBlockData to initialize data for all blocks even if we are only going to visit a few.

That could change the algorithmic behavior to change, no? Is it possible to make BasicBlockData initialize by default in a lazy way? That might help. But I am still worried about mallocing for all blocks in a function. @eeckstein your thoughts?

File: lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp (hunk 4 above)
Comment (gottesmm) [reply lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp:366]: I actually think the assert wouldn't be too difficult (consider a copy + sortUnique + ensure nothing removed), that being said I think such an assert would be something for a lower part of the API.

**Review by gottesmm — COMMENTED** [review]
Closer.

File: include/swift/SIL/SILFunction.h [hunk 12]
```diff
@@ -199,7 +199,7 @@ class SILFunction
 
   /// A monotonically increasing ID which is incremented whenever a
   /// BasicBlockBitfield is constructed.
-  /// Usually this stays below 1000, so a 32-bit unsigned is more than
+  /// Usually this stays below 100000, so a 32-bit unsigned is more than
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply include/swift/SIL/SILFunction.h:202]: There is already an assert in the BasicBlockBitfield constructor

File: lib/SIL/Utils/BasicBlockUtils.cpp (hunk 9 above)
Comment (gottesmm) [reply lib/SIL/Utils/BasicBlockUtils.cpp:428]: Erik and I spoke about this and he mentioned that the reason why he didn't do that is he was worried about algorithmic changes in how the algorithm worked in this case. I am going to think about that and if we change this it will be in a follow on commit.

File: lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp (hunk 11 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:407]: I talked with eeckstein offline and he said that he only changed to use BasicBlockData for cases where we have to visit the whole CFG, so I'm cool with this.

**Review by gottesmm — APPROVED** [review]
Just make the rest of the changes/etc. This LGTM!

File: lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp (hunk 11 above)
Comment (eeckstein) [reply lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:407]: In case of YieldOnceCheck the "visitedBBs" data is only created once a run of the pass. And BasicBlockData should only be used for such cases.
Yes, we could make an extended version of BasicBlockData which lazily constructs data (by using it together with BasicBlockFlag). That's future work.

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge
